### PR TITLE
Ostimove

### PR DESCRIPTION
**Pull request by robk5uj**

### CODE REVIEW

**Comment by robk5uj** [comment]
This should transfer the OSTI DOIs into the 035 field and eliminate the associated 856.

**Comment by tsgit** [comment]
please rebase so that only the osti changes are part of the pull request

File: bibcheck/plugins/ostimove.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                                     subfield_filter=('y', PROVENANCE)):
+        if val:
+            ostimatch = OSTI_ID.match(val)
+            if ostimatch:
+                ostimatch_tuple = ostimatch.groups()
+                ostimatch_string = ''.join(ostimatch_tuple)
+                osti_subfields = [('9', 'OSTI'), ('a', ostimatch_string)]
+                record_add_field(record, '035', ' ', ' ',
+                                     subfields=osti_subfields)
+                record.delete_field((pos[0][0:3], pos[1], None))
```
Comment (tsgit) [inline bibcheck/plugins/ostimove.py]: this will break when there is more than one field 8564 with OSTI content

why? because you remove from the record at a particular position and then you use the old position -- which now may not be there any longer or have other content to remove the second occurence

Comment (tsgit) [inline bibcheck/plugins/ostimove.py]: there's 2 ways to address this, a keep track of how many fields were deleted and adjust pos in the next call to delete_field
or work on the reverse list and remove from the end

**Comment by robk5uj** [comment]
Oh, okay

**Comment by tsgit** [comment]
here is the error I get when I run it:

```
2016-07-21 13:35:18 --> Amended record 772625 by rule ostimove: Deleted field 856
2016-07-21 13:35:18 --> Task #846425 finished. [RUNNING]
2016-07-21 13:35:18 --> Unexpected error occurred: pop index out of range.
2016-07-21 13:35:18 --> Traceback is:
2016-07-21 13:35:18 --> * 2016-07-21 13:35:18 -> IndexError: pop index out of range (bibcheck_task.py:446:delete_field)
2016-07-21 13:35:18 --> ** User details
2016-07-21 13:35:18 --> No client information available
2016-07-21 13:35:18 --> ** Traceback details 
2016-07-21 13:35:18 --> Traceback (most recent call last):
2016-07-21 13:35:18 -->   File "/scratch/venvs/invenio-legacy/lib/python2.7/site-packages/invenio/bibtask.py", line 620, in task_init
2016-07-21 13:35:18 -->     ret = _task_run(task_run_fnc)
2016-07-21 13:35:18 -->   File "/scratch/venvs/invenio-legacy/lib/python2.7/site-packages/invenio/bibtask.py", line 1184, in _task_run
2016-07-21 13:35:18 -->     if callable(task_run_fnc) and task_run_fnc():
2016-07-21 13:35:18 -->   File "/scratch/venvs/invenio-legacy/lib/python2.7/site-packages/invenio/bibcheck_task.py", line 606, in task_run_co
re
2016-07-21 13:35:18 -->     check_record(rule, record)
2016-07-21 13:35:18 -->   File "/scratch/venvs/invenio-legacy/lib/python2.7/site-packages/invenio/bibcheck_task.py", line 684, in check_recor
d
2016-07-21 13:35:18 -->     return plugin["check_record"](record, **rule["checker_params"])
2016-07-21 13:35:18 -->   File "/scratch/venvs/invenio-legacy/lib/python/invenio/bibcheck_plugins/ostimove.py", line 45, in check_record
2016-07-21 13:35:18 -->     record.delete_field((pos[0][0:3], pos[1], None))
2016-07-21 13:35:18 -->   File "/scratch/venvs/invenio-legacy/lib/python2.7/site-packages/invenio/bibcheck_task.py", line 446, in delete_fiel
d
2016-07-21 13:35:18 -->     return self._query(position[:1] + (None, None)).pop(position[1])
2016-07-21 13:35:18 --> IndexError: pop index out of range
2016-07-21 13:35:18 --> ** Stack frame details
2016-07-21 13:35:18 --> Frame delete_field in /scratch/venvs/invenio-legacy/lib/python2.7/site-packages/invenio/bibcheck_task.py at line 446
2016-07-21 13:35:18 --> -------------------------------------------------------------------------------
2016-07-21 13:35:18 -->        443             message = u"Deleted field %s" % (position[0])
2016-07-21 13:35:18 -->        444         self.set_amended(message)
2016-07-21 13:35:18 -->        445         if position[2] is None:
2016-07-21 13:35:18 --> ---->  446             return self._query(position[:1] + (None, None)).pop(position[1])
2016-07-21 13:35:18 -->        447         else:
2016-07-21 13:35:18 -->        448             return self._query(position[:2] + (None,))[0].pop(position[2])
2016-07-21 13:35:18 -->        449 
2016-07-21 13:35:18 --> -------------------------------------------------------------------------------
2016-07-21 13:35:18 -->                     position =  "('856', 4, None)"
2016-07-21 13:35:18 -->                      message =  "u'Deleted field 856'"
2016-07-21 13:35:18 -->                         self =  "{'595': [([('9', 'SPIRES-HIDDEN'), ('a', 'Added by ..HEP.ADD.TO.HEP from')], ' ', ' 
', '', 19), ([('9', 'SPIRES-HIDDEN'), ('a', 'TeXtitle from script')], ' ', ' ', '', 20)], '773': [([('c', '1191'), ('p', 'Conf.Proc.'), ('v',
 'C070625'), ('y', '2007')], ' ', ' ', '', 54), ([('c', '1191-1193'), ('w', 'C07-06-25'), ('y', '2007')], ' ', ' ', '', 55)], '961': [([('x',
 '2008-01-10')], ' ', ' ', '', 62), ([('c', '2011-12-16')], ' ', ' ', '', 63)], '300': [([('a', '3')], ' ', ' ', '', 16)], '980': [([ [...]
2016-07-21 13:35:18 --> Frame check_record in /scratch/venvs/invenio-legacy/lib/python/invenio/bibcheck_plugins/ostimove.py at line 45
2016-07-21 13:35:18 --> -------------------------------------------------------------------------------
```

**Comment by tsgit** [comment]
also, 
```
2016-07-21 13:34:42 --> Processing record 336606
2016-07-21 13:34:42 --> Entering task_sleep_now_if_required with status=RUNNING
2016-07-21 13:34:42 --> Amended record 336606 by rule ostimove: Deleted field 856
2016-07-21 13:34:42 --> Amended record 336606 by rule ostimove: 336606: moved link 'http://www.osti.gov/energycitations/product.biblio.jsp?osti_id=7271026'
2016-07-21 13:34:42 --> Amended record 336606 by rule ostimove: Deleted field 856
2016-07-21 13:34:42 --> Amended record 336606 by rule ostimove: 336606: moved link 'http://www.osti.gov/energycitations/product.biblio.jsp?osti_id=7272047'
2016-07-21 13:34:42 --> Updating task progress to Processing record 6847/12605 (54%)..
```

but that's not true, due to the shift in fields after the first one is deleted, the second deletion actually affects a valid URL.

the bibupload file has
```
  <datafield tag="856" ind1="4" ind2=" ">
    <subfield code="u">http://lss.fnal.gov/cgi-bin/find_paper.pl?other/ssc/sscl-preprint-132-Rev-1</subfield>
    <subfield code="w">ssc/sscl-preprint-132-Rev-1</subfield>
    <subfield code="y">FERMILABOTHER</subfield>
  </datafield>
  <datafield tag="856" ind1="4" ind2=" ">
    <subfield code="u">http://www.osti.gov/energycitations/product.biblio.jsp?osti_id=7272047</subfield>
    <subfield code="w">7272047</subfield>
    <subfield code="y">OSTI</subfield>
  </datafield>
```

while the original has

```
<datafield tag="856" ind1="4" ind2=" "><subfield code="u">http://lss.fnal.gov/cgi-bin/find_paper.pl?other/ssc/sscl-preprint-132-Rev-1</subfield><subfield code="w">ssc/sscl-preprint-132-Rev-1</subfield><subfield code="y">FERMILABOTHER</subfield></datafield><datafield tag="856" ind1="4" ind2=" "><subfield code="u">http://www.osti.gov/energycitations/product.biblio.jsp?osti_id=7271026</subfield><subfield code="w">7271026</subfield><subfield code="y">OSTI</subfield></datafield><datafield tag="856" ind1="4" ind2=" "><subfield code="u">http://www.osti.gov/energycitations/product.biblio.jsp?osti_id=7272047</subfield><subfield code="w">7272047</subfield><subfield code="y">OSTI</subfield></datafield><datafield tag="856" ind1="4" ind2=" "><subfield code="u">http://inspirehep.net/record/336606/files/sscl-preprint-132-.pdf</subfield></datafield>
```

File: bibcheck/plugins/ostimove.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+""" Bibcheck plugin to move info for OSTI from tag
+    8564_w/y to tag 035__a/9
+"""
+
+from invenio.bibrecord import record_add_field
+import re
+
+PROVENANCE = 'OSTI'
+
+OSTI_ID = re.compile(r'^https?://(?:www\.)?osti\.gov/.*?\D(\d+)$')
```
Comment (tsgit) [inline bibcheck/plugins/ostimove.py]: I find several osti links which do not match this regexp

e.g.
http://www.osti.gov/accomplishments/documents/fullText/ACC0169.pdf                                                                   
http://www.osti.gov/scitech/servlets/purl/5896179/                                                                                       
http://www.osti.gov/accomplishments/documents/fullText/ACC0068.pdf                                                                         
http://www.osti.gov/energycitations/product.biblio.jsp?osti_id=5971122'DOC-TYPE=ConferencePaper                                               
http://www.osti.gov/scitech/servlets/purl/1000476-RU4rvP/                                                                                         
http://www.osti.gov/scitech/servlets/purl/7250813-qyOwlS/                                                                                          
http://www.osti.gov/scitech/servlets/purl/6960980-5T3xqy/                                                                                                
http://www.osti.gov/scitech/servlets/purl/5473655-sQMxI2/                                                                                                  
http://www.osti.gov/scitech/biblio/10117897-from-psi-charmed-mesons
http://www.osti.gov/energycitations/product.biblio.jsp?osti_id=
http://www.osti.gov/energycitations/product.biblio.jsp?osti_id=
http://www.osti.gov/bridge/servlets/purl/495783-khmhVM/webviewable/495783.pdf
http://www.osti.gov/scitech/servlets/purl/789443/
http://www.osti.gov/scitech/servlets/purl/981461/
http://www.osti.gov/accomplishments/documents/fullText/ACC0067.pdf
http://www.osti.gov/bridge/purl.cover.jsp?purl=/4096074/4096074.pdf
http://www.osti.gov/scitech/servlets/purl/1133081/
http://www.osti.gov/scitech/servlets/purl/472669-2yrzjX/webviewable/
http://www.osti.gov/scitech/servlets/purl/654116-I7Z6F1/webviewable/

Comment (tsgit) [inline bibcheck/plugins/ostimove.py]: also, I don't find a single osti URL which doesn't start with http://www. ,
so why is that made optional in the regexp?

File: bibcheck/plugins/ostimove.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    """ move 8564_u/y to 035__a/9 """
+    for pos, val in record.iterfield('8564_u',
+                                     subfield_filter=('y', PROVENANCE)):
+        if val:
+            ostimatch = OSTI_ID.match(val)
+            if ostimatch:
+                ostimatch_tuple = ostimatch.groups()
+                ostimatch_string = ''.join(ostimatch_tuple)
+                osti_subfields = [('9', 'OSTI'), ('a', ostimatch_string)]
+                record_add_field(record, '035', ' ', ' ',
```
Comment (tsgit) [inline bibcheck/plugins/ostimove.py]: there are no precautions here against adding the same info again if there are repeated marc tags 8564. this would be a perfect opportunity to avoid redundant information. the adsmove script specifically addressed these issue -- see e.g. https://inspirehep.net/record/601165/export/hm

File: bibcheck/plugins/ostimove.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+
+def check_record(record):
+    """ move 8564_u/y to 035__a/9 """
+    for pos, val in record.iterfield('8564_u',
+                                     subfield_filter=('y', PROVENANCE)):
+        if val:
+            ostimatch = OSTI_ID.match(val)
+            if ostimatch:
+                ostimatch_tuple = ostimatch.groups()
```
Comment (tsgit) [inline bibcheck/plugins/ostimove.py]: I do not understand why you cast this into a tuple to join. ostimatch.group(1) is more succinct, and avoids unnecessary and confusing hoops. why was this changed from adsmove?

File: bibcheck/plugins/ostimove.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                ostimatch_string = ''.join(ostimatch_tuple)
+                osti_subfields = [('9', 'OSTI'), ('a', ostimatch_string)]
+                record_add_field(record, '035', ' ', ' ',
+                                     subfields=osti_subfields)
+                record.delete_field((pos[0][0:3], pos[1], None))
+                record.set_amended(
+                    "%s: moved link '%s'" % (record.record_id, val))
+            else:
+                print "%s: no match for '%s'" % (record.record_id, val)
```
Comment (tsgit) [inline bibcheck/plugins/ostimove.py]: print statements are to be avoided, because you don't know where stdout is going -- you should use record.warn() instead

File: bibcheck/plugins/ostimove.py (hunk 1 above)
Comment (tsgit) [inline bibcheck/plugins/ostimove.py]: in many cases 8564 also contains $$w 
this deletes the entire tag without checking that the info in $$w agreed with the info in $$u   they should have the same osti identifiers
otherwise you throw away potentially useful information

Comment (tsgit) [inline bibcheck/plugins/ostimove.py]: I performed a quick and dirty check
```

In [11]: for r in perform_request_search(p="8564_y:osti"):                                                                                                                                               
    fv = get_fieldvalues(r, '8564_u')                                                                                                                                   
    fv = [ore.match(u).group(1) for u in fv if ore.match(u)]                                                                                                            
   ....:     if len(fv) == 1:                                                                                                                                           
   ....:         ws = get_fieldvalues(r, '8564_w')
   ....:         if ws and fv[0] not in ws:                                                                                                                             
   ....:             print r, fv, ws
   ....:             
127467 ['6534741'] ['p7803945']
256548 ['10105486'] ['4252']
417505 ['228486'] ['9921']
463206 ['658361'] ['441/289']
570177 ['789433'] ['p01/PAPERS/RPPH048.PDF']
570194 ['789381'] ['p01/PAPERS/RPPH074.PDF']
848315 ['941166'] ['074/079/Nufact08_079.pdf']
1302269 ['1110727'] ['COOL2013/papers/moam2ha04.pdf']
1302271 ['1113193'] ['PAC2013/papers/thpho05.pdf']
1302275 ['1113191'] ['PAC2013/papers/thpho13.pdf']
1302752 ['1113188'] ['PAC2013/papers/thpba08.pdf']
```
those are all because there are other 8564 tags with $$w, none of them indicate a conflict

**Comment by tsgit** [comment]
closed, superseeded by #3c62ae0df2939056fd1bf3c6f6540c1030c7bb74
